smithy-model/src/main/java/software/amazon/smithy/model/node/IdentityClassCache.java

```diff
@@ -0,0 +1,54 @@
+/*
+ * Copyright 2022 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+package software.amazon.smithy.model.node;
+
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ConcurrentMap;
+import java.util.function.Supplier;
+
+/**
+ * Caches instances of type T associated with given a Class.
+ *
+ * <p>Cached values are invalidated if the requested class != currently cached
+ * class. This ensures that the most recently loaded versions of classes are
+ * being used.
+ *
+ * @param <T> The type of the cached classes.
+ */
+final class IdentityClassCache<T> {
+
+    private final ConcurrentMap<String, T> cache = new ConcurrentHashMap<>();
+
+    /**
+     * Retrieves cached instance of T associated with the given Class,
+     * computing the value to store if the given Class has no associated
+     * value, or if the cache is invalidated.
+     *
+     * @param forClass The exact Class of value to retrieve from the cache.
+     * @param supplierIfNotPresent Function to compute cached value.
+     * @return Cached value.
+     */
+    T get(Class<?> forClass, Supplier<? extends T> supplierIfNotPresent) {
```

Comment:
```suggestion
    T get(Class<? extends T> forClass, Supplier<? extends T> supplierIfNotPresent) {
```
I think we'd need something like this to make sure comparison to invalidate the cache works properly.

Reply:
I don't think this will address your observation that the class key type != the cached return value. Instead, maybe track the original cache key in a Pair and then check the original cache key class vs the given class to invalidate the cache. Something like:

```
final class IdentityClassCache<T> {

    private final ConcurrentMap<String, Pair<Class<?>, T>> cache = new ConcurrentHashMap<>();

    T get(Class<?> forClass, Supplier<? extends T> supplierIfNotPresent) {
        return cache.compute(forClass.getName(), (name, current) -> {
            if (current == null || current.getLeft() != forClass) {
                return Pair.of(forClass, supplierIfNotPresent.get());
            } else {
                return current;
            }
        }).getRight();
    }
}
```

Reply:
Yea this makes sense. My only question is, do we want to change the keys that are currently being used in [DefaultNodeDeserializers:626](https://github.com/awslabs/smithy/pull/1518/files#diff-b97479d8946922af704f3004a94c1e7f7e147ff551690953364c254cf6665b16L626)? It uses the node class *and* the target type to create a key, whereas these ideas would just be using the target type as a key. Is there a situation where a single node class may have multiple target types, with different ObjectCreators?